tests/cpp/test_allocation_domain.cpp

```diff
@@ -1414,4 +1417,50 @@ TEST_F(AllocationDomainTest, ClearReductionIterDomainsPatch) {
       tv1->getContiguity(), ElementsAre(contig_copy[0], contig_copy[2]));
 }
 
+TEST_F(AllocationDomainTest, InputAllocationIsSplit_Concrete) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  TensorView* in = makeContigConcreteTensor({6});
+  TensorView* out = set(in);
+  in->split(0, 2);
+  in->setAllocationDomain(in->getLoopDomain(), true);
+
+  fusion->addInput(in);
+  fusion->addOutput(out);
+
+  FusionExecutorCache executor_cache(std::move(fusion));
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA);
+  at::Tensor in_tensor = at::randn({6}, options);
+  auto out_tensors = executor_cache.runFusionWithInputs({in_tensor});
+
+  testValidate(
+      executor_cache.fusion(), out_tensors, {in_tensor}, __LINE__, __FILE__);
+}
+
+// The test fails as is. The symbolic IterDomains in loop/allocation are not
+// concretized. I tried to change DynamicTransformConcretizer::mutate to grab
+// all expressions between root and allocation but still couldn't get it to
+// work.
```

Comment:
@jacobhinkle FYI

Reply:
I'm not sure why this could cause that problem. It may be just because of the concern I mentioned below.

Reply:
This issue persisted after I moved the split after addInput/addOutput. FYI, below is my failed attempt to fix dynamic transform for this test. 

```
diff --git a/csrc/dynamic_transform.cpp b/csrc/dynamic_transform.cpp
index 24404db8..d287149d 100644
--- a/csrc/dynamic_transform.cpp
+++ b/csrc/dynamic_transform.cpp
@@ -1056,7 +1056,7 @@ void DynamicTransformConcretizer::mutate(TensorView* tv) {
     // beyond the logical domain as asserted above
     auto all_id_exprs = StmtSort::getExprsBetween(
         {tv->getRootDomain().begin(), tv->getRootDomain().end()},
-        {tv->getLogicalDomain().begin(), tv->getLogicalDomain().end()});
+        {tv->getMaybeAllocationDomain().begin(), tv->getMaybeAllocationDomain().end()});
     for (auto expr : all_id_exprs) {
       // Assume outputs of IterDomain exprs are always IterDomains. If
       // the assumption is invalidated, the logic here would need to
```

Reply:
Previously we had assumed there were no loop or allocation transforms so that each of those would just be permutations of logical. I think what we should do is actually use IRBFS here to propagate from root to all of the other domains (logical, loop, and allocation).

Reply:
Did you try it? While I agree with what you said, I doubt that helps this particular test case where root->maybeallocation includes all expressions. Btw, I believe https://github.com/NVIDIA/Fuser/blob/b5e51824cacd910142ed01635a4453bb5070ccd8/csrc/ir/nodes.cpp#L3762 can be used to capture all Exprs in a TensorView.

Reply:
I did not try it yet. I looked at the code again though and I see. I think we only inspect logical IDs for concretizing IterType in DynamicTransformInitialInfo. We just need to check out the allocation and loop domains also when building that in case they are set. I'll have a look in the AM.

Re allExprs(), yes but they need to be topologically sorted to do the propagation at concretization.

Reply:
> I'll have a look in the AM.

Thanks! 

> Re allExprs(), yes but they need to be topologically sorted to do the propagation at concretization.

FYI, I believe it's topo-sorted already: https://github.com/NVIDIA/Fuser/blob/b5e51824cacd910142ed01635a4453bb5070ccd8/csrc/ir/nodes.cpp#L3759

Reply:
Ah, you're right.

Reply:
This fusion is actually barely dynamic:
```
Fusion before concretization:
Inputs:
  T0_g_float[ iS2{( ceilDiv(i0, 2) )}, iS3{2} ]
Outputs:
  T1_g_float[ iS1{i0} ]

%kernel_math {
T1_g_float[ iS1{i0} ]
   = Set( T0_g_float[ iS2{( ceilDiv(i0, 2) )}, iS3{2} ], cache_op=Streaming )
} // %kernel_math

DynamicTransformInitialInfo
  Dynamic reshaped TensorViews:
  Dynamic resized IterDomains:
  Dynamic expanded TensorViews:
  Dynamic factory-function output TensorViews:
  Dynamic extent Vals:
    i0
  Root dynamic Vals:
    i0

DynamicTransformConcretizationInfo
  Empty tensor extents:
  Reshape:
  Resize:
  Expand:
  Factory Output IterTypes:

Concretized Fusion:

%kernel {
T1_g_float[ iS1{i0} ]
   = Set( T0_g_float[ iS2{( ceilDiv(i0, 2) )}, iS3{2} ], cache_op=Streaming )

TransformPrinter :
T0_g_float[ iS2{( ceilDiv(i0, 2) )}, iS3{2} ]
 logical domain : (iS0{i0})
 allocation domain : (iS2{( ceilDiv(i0, 2) )}, iS3{2})
 contiguity: t t
  Split: iS0{i0} by factor 2 -> iS2{( ceilDiv(i0, 2) )}, iS3{2}
 loop domain : (iS2{( ceilDiv(i0, 2) )}, iS3{2})
T1_g_float[ iS1{i0} ]
 logical domain : (iS1{i0})
 contiguity: t
 loop domain : (iS1{i0})
} // %kernel
```
There are no `IterType::Symbolic` IDs in the unconcretized fusion. The only thing that's happening at concretization is we are checking whether `i0` is 0, so the fusion is untouched by concretization.

The error I am seeing is actually in the `replaceSymbolicSizes` pass. When we try to replace `i0` there with `( (( (( getMetaData(T0) )).logical_size ))[0] )`, we hit this error:
```
C++ exception with description " INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/ir/utils.cpp":965, please report a bug with repro script to NVFuser at https://github.com/NVID
IA/Fuser/issues. dom0 has unreachable IDs. dom0: iS27{( (( (( getMetaData(T0) )).logical_size ))[0] )}. dom1: iS2{( ceilDiv(i0, 2) )}, iS3{2}                                       
Exception raised from validateDomainEquivalence at /opt/pytorch/nvfuser/csrc/ir/utils.cpp:965 (most recent call first): 
```
So the error happens in`ir_utils::replaceValue` which uses `OptOutMutator`. This uses `StmtSort` to get statements. In particular that class uses `MemberStatements` when `traverse_members=true`. I originally thought the issue might lie there, at https://github.com/NVIDIA/Fuser/blob/769a4d24f92f9bf0d404bb3643dde15b07770d5e/csrc/iter_visitor.cpp#L71C10-L74C4
This is getting all loop IDs instead of all "leaf IDs", but that is not the problem in this case.

In this input TV, the logical domain is processed but not the loop/allocation:
```
165       Val* mutated_val = IrBuilder::createInContainer<TensorDomain>(
166           td->container(),
167           root_dom,
168           logical_dom,
169           allocation_dom,
(gdb) pv logical_dom
iS27{( (( (( getMetaData(T0) )).logical_size ))[0] )}
(gdb) pv allocation_dom
iS2{( ceilDiv(i0, 2) )}
iS3{2}
```
This shows that the logical domain is already mutated at this point but `iS2` and `iS3` should also be replaced since they're siblings and `iS2`'s extent contains a mutated Val.

So it seems the actual problem is that `StmtSort::traverseBetween` is not traversing anything but the logical domain of fusion input TVs. When `traverse_members=true`, we should probably be traversing `tv->domain()->allIDs()` for each TensorView including the inputs.

Reply:
Yes, the issue here is that allocation and logical and disconnected because there's been a replacement performed on logical, but not on allocation. I think the issue is that `StmtSort::getStmts` is only giving us the logical domains of the input TVs whereas we should expect it to provide all IDs for processing before the input TensorDomain and the TV itself.

Reply:
Ha, looks like it's actually just loop domains.

https://github.com/NVIDIA/Fuser/blob/main/csrc/iter_visitor.cpp#L73

We should probably change this to visit all of root, logical and loop domains.

Reply:
In this particular case though, loop is equal to allocation. I originally thought this was the issue here. I agree that that line is an issue if we have unconventional tensor domains like loops that are producers of logical, and we should ensure all the domains are available there, but I think in this case the bad behavior is specific to input TVs.

Reply:
Not for this PR. I can stick with concrete sizes for the tests for now